cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java

```diff
@@ -77,62 +78,154 @@ public LogCleanup(FileMetaDataManager fileMetaDataManager, RootLocationFactory r
+   *
+   * @param tillTime                time till the meta data will be deleted.
+   * @param batchSize               max number files to be deleted in one iteration of log cleanup
+   * @param namespacedLogBaseDirMap namespace to directory map
+   * @param parentDirs              parent directories for deleted files
+   * @return map of location and namespace
+   */
+  @VisibleForTesting
+  Map<URI, NamespaceId> cleanupFileLocations(final long tillTime, final int batchSize,
+                                             final Map<String, NamespaceId> namespacedLogBaseDirMap,
+                                             final SetMultimap<String, Location> parentDirs) {
+    int count = 0;
+    FileMetaDataManager.TableKey nextTableKey = null;
+    Map<URI, NamespaceId> metaFilesToDelete = new HashMap<>();
+
+    do {
+      // Get the log metadata in batches
+      FileMetaDataManager.MetaFileProcessor<Map<URI, NamespaceId>> metaFileProcessor =
+        new FileMetaDataManager.MetaFileProcessor<Map<URI, NamespaceId>>() {
+          Map<URI, NamespaceId> scannedFiles = new HashMap<>();
+
+          @Override
+          public void process(URI uri, NamespaceId namespaceId) {
+            scannedFiles.put(uri, namespaceId);
+          }
+
+          @Override
+          public Map<URI, NamespaceId> getCollectedFiles() {
+            return scannedFiles;
+          }
+        };
+
+      nextTableKey = fileMetaDataManager.scanFiles(nextTableKey, batchSize, metaFileProcessor);
+      Map<URI, NamespaceId> scannedMetaFiles = metaFileProcessor.getCollectedFiles();
+
+      for (Map.Entry<URI, NamespaceId> entry : scannedMetaFiles.entrySet()) {
+        final URI file = entry.getKey();
+
+        try {
+          Location fileLocation = impersonator.doAs(entry.getValue(), new Callable<Location>() {
+            @Override
+            public Location call() throws Exception {
+              return rootLocationFactory.create(file);
+            }
+          });
+
+          if (!fileLocation.exists()) {
+            LOG.warn("Log file {} does not exist, but metadata is present", fileLocation.toString());
+            metaFilesToDelete.put(entry.getKey(), entry.getValue());
+            count++;
+          } else if (fileLocation.lastModified() < tillTime) {
+            deleteLogFiles(parentDirs, namespacedLogBaseDirMap, entry.getValue(), fileLocation);
+            metaFilesToDelete.put(entry.getKey(), entry.getValue());
+            count++;
+          }
+        } catch (Exception e) {
+          if (e instanceof NamespaceNotFoundException) {
+            LOG.warn("Namespace does not exist for {}. Going to delete metadata for it", file.toString(), e);
+            metaFilesToDelete.put(entry.getKey(), entry.getValue());
+            count++;
+          } else {
+            LOG.warn("Got exception while accessing path {}", file.toString(), e);
+          }
+        }
+      }
+    } while (nextTableKey != null && count != batchSize);
```

Comment:
yes, we do not need both the conditions. We can only keep `nextTableKey != null`